trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,251 @@
+        """
+        # we are recovering previously saved optimization state
+        # so the model was already trained
+        # thus there is no need to train it again
+        return cls(
+            search_space,
+            record.datasets,
+            record.models,
+            acquisition_rule=acquisition_rule,
+            acquisition_state=record.acquisition_state,
+            fit_model=False,
+        )
+
+    def to_record(self) -> Record[S]:
+        """Collects the current state of the optimization, which includes datasets,
+        models and acquisition state (if applicable).
+
+        :return: An optimization state record.
+        """
+        return Record(
+            datasets=self._datasets, models=self._models, acquisition_state=self._acquisition_state
+        )
+
+    def to_result(self) -> OptimizationResult[S]:
+        """Converts current state of the optimization
+        into an :class:`~trieste.data.OptimizationResult` object."""
+        record: Record[S] = self.to_record()
+        return OptimizationResult(Ok(record), [])
+
+    def ask(self) -> TensorType:
+        """Suggests a point (or points in batch mode) to observe by optimizing the acquisition function.
+        If the acquisition is stateful, its state is saved.
+
+        :return: A :class:`TensorType` instance representing suggested point(s).
+        """
+        # This trick deserves a comment to explain what's going on
+        # acquisition_rule.acquire can return different things:
+        # - when acquisition has no state attached, it returns just points
+        # - when acquisition has state, it returns a Callable
+        #   which, when called, returns state and points
+        # so code below is needed to cater for both cases
+
+        points_or_stateful = self._acquisition_rule.acquire(
+            self._search_space, self._datasets, self._models
+        )
+
+        if callable(points_or_stateful):
+            self._acquisition_state, query_points = points_or_stateful(self._acquisition_state)
+        else:
+            query_points = points_or_stateful
+
+        return query_points
+
+    def tell(self, new_data: Mapping[str, Dataset] | Dataset) -> None:
+        """Updates optimizer state with new data.
+
+        :param new_data: New observed data.
+        :raise ValueError: If keys in ``new_data`` do not match those in already built dataset.
+        """
+        if isinstance(new_data, Dataset):
+            new_data = {OBJECTIVE: new_data}
+
+        if self._datasets.keys() != new_data.keys():
+            raise ValueError(
+                f"new_data keys {new_data.keys()} doesn't "
+                f"match dataset keys {self._datasets.keys()}"
+            )
+
+        self._datasets = {tag: self._datasets[tag] + new_data[tag] for tag in new_data}
+
+        for tag, model in self._models.items():
+            dataset = self._datasets[tag]
+            model.update(dataset)
+            model.optimize(dataset)
```

Comment:
I wonder if you always want to optimize the model after every tell?

Reply:
It might be a useful idea, although I am struggling to come up with a concrete example. It is easy to add it later if needed, so let's leave it like this for now, and see if relevant use case comes up. Sounds good?

Reply:
If the model fitting is slow, you might want to do a few ask-tells before refitting. But yeah, lets leave it fo rnow